flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -833,10 +832,44 @@ public TableResultInternal executeInternal(List<ModifyOperation> operations) {
         return executeInternal(transformations, sinkIdentifierNames, jobStatusHookList);
     }
 
-    private ModifyOperation getOperation(ReplaceTableAsOperation rtasOperation) {
-        // rtas drop table first, then create
+    private ModifyOperation getModifyOperation(
+            ReplaceTableAsOperation rtasOperation, List<JobStatusHook> jobStatusHookList) {
         CreateTableOperation createTableOperation = rtasOperation.getCreateTableOperation();
         ObjectIdentifier tableIdentifier = createTableOperation.getTableIdentifier();
+        Catalog catalog = catalogManager.getCatalog(tableIdentifier.getCatalogName()).orElse(null);
+        ResolvedCatalogTable catalogTable =
+                catalogManager.resolveCatalogTable(createTableOperation.getCatalogTable());
+        Optional<DynamicTableSink> stagingDynamicTableSink =
+                getSupportsStagingDynamicTableSink(createTableOperation, catalog, catalogTable);
+        if (stagingDynamicTableSink.isPresent()) {
+            // use atomic rtas
+            DynamicTableSink dynamicTableSink = stagingDynamicTableSink.get();
+            if (!rtasOperation.isCreateOrReplace()) {
+                Optional<ContextResolvedTable> driverTable =
+                        catalogManager.getTable(tableIdentifier);
+                if (!driverTable.isPresent()) {
+                    throw new TableException(
+                            String.format(
+                                    "The table %s to be replaced doesn't exist. "
+                                            + "You can try to use CREATE TABLE AS statement or "
+                                            + "CREATE OR REPLACE TABLE AS statement.",
+                                    tableIdentifier));
+                }
+            }
+            SupportsStaging.StagingPurpose stagingPurpose =
+                    rtasOperation.isCreateOrReplace()
+                            ? SupportsStaging.StagingPurpose.CREATE_OR_REPLACE_TABLE_AS
+                            : SupportsStaging.StagingPurpose.REPLACE_TABLE_AS;
+
+            StagedTable stagedTable =
+                    ((SupportsStaging) dynamicTableSink)
+                            .applyStaging(new SinkStagingContext(stagingPurpose));
+            AtomicJobStatusHook atomicJobStatusHook = new AtomicJobStatusHook(stagedTable);
+            jobStatusHookList.add(atomicJobStatusHook);
+            return rtasOperation.toStagedSinkModifyOperation(
+                    tableIdentifier, catalogTable, catalog, dynamicTableSink);
+        }
+        // non-atomic rtas drop table first, then create
         try {
             catalogManager.dropTable(tableIdentifier, rtasOperation.isCreateOrReplace());
         } catch (ValidationException e) {
```

Comment:
Just notice we will use error message to check whether the table to be dropped exists or not. For me, it's hacky and when the error message changes, the check may well fail.
Can we first get the table and then drop?